### DISCUSSION TITLE
Rate-Limiting Policy

### DISCUSSION THREAD

**Discussion by Paul2708** [open-policy-agent/community#611] [OPA and Rego] [answered]
Hey there!

I am aware that, for example, Envoy provides a [rate-limiting filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/rate_limit_filter).
However, I am curious if it is possible to implement a rate-limiting policy in pure Rego somehow.
There is a GitHub issue https://github.com/open-policy-agent/opa/issues/632 mentioning a rate-limiting policy. However, I could not find it in the referred pull request.
Also, I did not find a way to modify an object to keep track of the sent requests.

Has someone already implemented a Rego policy for rate limiting, or is it not possible with OPA and Rego?

Thanks in advance!

Best regards,
Paul

**Comment by srenatus** [comment, marked as answer]
> However, I am curious if it is possible to implement a rate-limiting policy in pure Rego somehow.

No, not with OPA alone. OPA doesn't keep state like it would be required to create a rate-limiting policy. You would need new builtins, or another process to update the OPA store with the required data.

**Reply by Paul2708** [reply]
Alright, thank you for clarifying it!
